README.md
# JpVaccination
It provides data on infant vaccination in Japan. It also processes calculations of recommended and next available dates for vaccination.

These data are based on [the vaccination schedule of the National Institute of Infectious Diseases](https://www.niid.go.jp/niid/ja/schedule.html). Currently, only regular and some optional vaccinations in infancy are covered, but in the future all vaccinations will be covered.

## Installation

Add this line to your application's Gemfile:

```ruby
gem 'jp_vaccination'
```

And then execute:

    $ bundle install

Or install it yourself as:

    $ gem install jp_vaccination

## Usage

```ruby
require 'JpVaccination'
```
### Access to vaccination data.
Pass the argument as a string for the [vaccination_key](#about-vaccination_key).
```ruby
chickenpox_1st = JpVaccination.find('chickenpox_1')

chickenpox_1st.name  "水痘"
chickenpox_1st.period "１回目"
chickenpox_1st.regular true
chickenpox_1st.type 生ワクチン
chickenpox_1st.recommended {:month=>12}
chickenpox_1st.deadline {:date_type=>"month", :start=>12, :end=>15, :less_than=>true}
chickenpox_1st.interval nil
```
Example data.
|column|data|example|title|
|-|-|-|-|
name |String|"水痘"|予防接種名|
period|String|"２回目"|期間|
regular|true/false|true|定期/任意|
type[^1]|String|"生ワクチン"|種類|
recommended|Hash|{ month: 18 }| 推奨接種月齢
deadline[^2]|Hash|{<br> date_type: "month",<br>start: 15,<br> end: 36,<br> less_than: true<br>}|接種期日|
interval[^2]|Hash|{ <br>date_type: "month",<br> start: 3,<br> end: 12 <br>}|接種間隔|

[^1]:`nil except for live vaccines.`
[^2]:5~36 months:{date_type:"month",start:15,end:36},less_than:true` is less than.
`less_than:false` is below.

### Calculate all recommended dates of vaccination
```rb
birthday = '2022-03-01'
pp JpVaccination.recommended_schedules(birthday, convert_to_strings = true)
# default convert_to_string: nil

=> [{:name=>"ヒブ １回目", :date=>"2022-05-01"},
    {:name=>"ヒブ ２回目", :date=>"2022-06-01"},
    {:name=>"ヒブ ３回目", :date=>"2022-07-01"},
    {:name=>"ヒブ ４回目", :date=>"2023-03-01"},
    {:name=>"Ｂ型肝炎 １回目", :date=>"2022-05-01"},
    {:name=>"Ｂ型肝炎 ２回目", :date=>"2022-06-01"},
    {:name=>"Ｂ型肝炎 ３回目", :date=>"2022-10-01"},
    {:name=>"ロタウイルス １回目", :date=>"2022-05-01"},
    {:name=>"ロタウイルス ２回目", :date=>"2022-06-01"},
    {:name=>"ロタウイルス ３回目", :date=>"2022-07-01"},
    {:name=>"小児用肺炎球菌 １回目", :date=>"2022-05-01"},
    {:name=>"小児用肺炎球菌 ２回目", :date=>"2022-06-01"},
    {:name=>"小児用肺炎球菌 ３回目", :date=>"2022-07-01"},
    {:name=>"小児用肺炎球菌 ４回目", :date=>"2023-03-01"},
    {:name=>"４種混合 第１期 １回目", :date=>"2022-06-01"},
    {:name=>"４種混合 第１期 ２回目", :date=>"2022-07-01"},
    {:name=>"４種混合 第１期 ３回目", :date=>"2022-08-01"},
    {:name=>"４種混合 第１期 ４回目", :date=>"2023-03-01"},
    {:name=>"２種混合 第２期", :date=>"2033-03-01"},
    {:name=>"ＢＣＧ ", :date=>"2022-08-01"},
    {:name=>"麻しん・風しん混合 第１期", :date=>"2023-03-01"},
    {:name=>"麻しん・風しん混合 第２期", :date=>"2027-04-01〜2028-04-01"},
    {:name=>"水痘 １回目", :date=>"2023-03-01"},
    {:name=>"水痘 ２回目", :date=>"2023-09-01"},
    {:name=>"おたふくかぜ １回目", :date=>"2023-03-01"},
    {:name=>"おたふくかぜ ２回目", :date=>"2027-04-01〜2028-04-01"},
    {:name=>"日本脳炎 第１期 １回目", :date=>"2025-03-01"},
    {:name=>"日本脳炎 第１期 ２回目", :date=>"2025-04-01"},
    {:name=>"日本脳炎 第１期 ３回目", :date=>"2026-03-01"},
    {:name=>"日本脳炎 第２期", :date=>"2031-03-01"}]
```
### Sort recommended vaccination dates in ascending order
```rb
birthday = '2022-03-01'
pp JpVaccination.sort_recommended_schedules(birthday, convert_to_strings = true)
# default convert_to_string: nil

=> {"2022-05-01"=>["ヒブ １回目", "ロタウイルス １回目", "小児用肺炎球菌 １回目", "Ｂ型肝炎 １回目"],
    "2022-06-01"=>["ヒブ ２回目", "ロタウイルス ２回目", "小児用肺炎球菌 ２回目", "４種混合 第１期 １回目", "Ｂ型肝炎 ２回目"],
    "2022-07-01"=>["ヒブ ３回目", "ロタウイルス ３回目", "小児用肺炎球菌 ３回目", "４種混合 第１期 ２回目"],
    "2022-08-01"=>["４種混合 第１期 ３回目", "ＢＣＧ "],
    "2022-10-01"=>["Ｂ型肝炎 ３回目"],
    "2023-03-01"=>["おたふくかぜ １回目", "ヒブ ４回目", "小児用肺炎球菌 ４回目", "水痘 １回目", "麻しん・風しん混合 第１期", "４種混合 第１期 ４回目"],
    "2023-09-01"=>["水痘 ２回目"],
    "2025-03-01"=>["日本脳炎 第１期 １回目"],
    "2025-04-01"=>["日本脳炎 第１期 ２回目"],
    "2026-03-01"=>["日本脳炎 第１期 ３回目"],
    "2027-04-01〜2028-03-31"=>["おたふくかぜ ２回目", "麻しん・風しん混合 第２期"],
    "2031-03-01"=>["日本脳炎 第２期"],
    "2033-03-01"=>["２種混合 第２期"]}
```

```ruby
pp JpVaccination.sort_recommended_schedules('2022-03-01')
=> {#<Date: 2022-05-01 ((2459701j,0s,0n),+0s,2299161j)>=>["小児用肺炎球菌 １回目", "ヒブ １回目", "ロタウイルス １回目", "Ｂ型肝炎 １回目"],…
```

### Calculate the next available vaccination date from the date of the previous vaccination.
```rb
pp JpVaccination.next_day(vaccination_key:'hepatitis_B_2', last_time: '2020-04-01')

=> {:name=>"Ｂ型肝炎 ２回目", :date=>#<Date: 2020-04-28 ((2458968j,0s,0n),+0s,2299161j)>}
```

If you want to find the date of the first vaccination, in short, the last character of the vaccination_key is 1, enter the date of birth in last_time.
```ruby
birthday = '2022-02-01'
pp JpVaccination.next_day(vaccination_key:'hepatitis_B_1', last_time: birthday)

=> {:name=>"Ｂ型肝炎 １回目", :date=>#<Date: 2022-04-01 ((2458968j,0s,0n),+0s,2299161j)>}
```

### Argument Error
If the [vaccination_key](#about-vaccination_key) is incorrect, the following error occurs.
```
# hepatitis_B_4 is incorrect.

puts JpVaccination.next_day(vaccination_key: 'hepatitis_B_4', last_time: '2020-04-01')

jp_vaccination/lib/jp_vaccination.rb:99:in `output_argument_error': The vaccination_key doesn't exist. (ArgumentError)
```

## About vaccination_key
The key formation is `(vaccination's English translation)_(period-number)`.

Each vaccination and corresponding key is as follows.

|vaccination_key|name|period|
|---|-----|------|
|hib_1|ヒブ|１回目|
|hib_2|ヒブ|２回目|
|hib_3|ヒブ|３回目|
|hib_4|ヒブ|４回目|
|hepatitis_B_1|Ｂ型肝炎|１回目|
|hepatitis_B_2|Ｂ型肝炎|２回目|
|hepatitis_B_3|Ｂ型肝炎|３回目|
|rotavirus_1|ロタウイルス|１回目|
|rotavirus_2|ロタウイルス|２回目|
|rotavirus_3|ロタウイルス|３回目|
|pneumococcus_1|小児用肺炎球菌|１回目|
|pneumococcus_2|小児用肺炎球菌|２回目|
|pneumococcus_3|小児用肺炎球菌|３回目|
|pneumococcus_4|小児用肺炎球菌|４回目|
|DPT_IPV_1|４種混合|第１期 １回目|
|DPT_IPV_2|４種混合|第１期 ２回目|
|DPT_IPV_3|４種混合|第１期 ３回目|
|DPT_IPV_4|４種混合|第２期|
|DT_1|２種混合|第２期|
|BCG_1|ＢＣＧ|-|
|MR_1|麻しん・風しん混合|第１期|
|MR_2|麻しん・風しん混合|第２期|
|chickenpox_1|水痘|１回目|
|chickenpox_2|水痘|２回目|
|mumps_1|おたふくかぜ|１回目|
|mumps_2|おたふくかぜ|２回目|
|Japanese_encephalitis_1|日本脳炎|第１期 １回目|
|Japanese_encephalitis_2|日本脳炎|第１期 ２回目|
|Japanese_encephalitis_3|日本脳炎|第１期 ３回目|
|Japanese_encephalitis_4|日本脳炎|第２期|

## Development

After checking out the repo, run `bin/setup` to install dependencies. Then, run `rake test` to run the tests. You can also run `bin/console` for an interactive prompt that will allow you to experiment.

To install this gem onto your local machine, run `bundle exec rake install`. To release a new version, update the version number in `version.rb`, and then run `bundle exec rake release`, which will create a git tag for the version, push git commits and the created tag, and push the `.gem` file to [rubygems.org](https://rubygems.org).

## Contributing

Bug reports and pull requests are welcome on GitHub at https://github.com/napple29/jp_vaccination. This project is intended to be a safe, welcoming space for collaboration, and contributors are expected to adhere to the [code of conduct](https://github.com/napple29/jp_vaccination/blob/master/CODE_OF_CONDUCT.md).

## License

The gem is available as open source under the terms of the [MIT License](https://opensource.org/licenses/MIT).

## Code of Conduct

Everyone interacting in the JpVaccination project's codebases, issue trackers, chat rooms and mailing lists is expected to follow the [code of conduct](https://github.com/napple29/jp_vaccination/blob/master/CODE_OF_CONDUCT.md).


CHANGELOG.md
## [Unreleased]

## [0.1.0] - 2022-03-07

- Initial release


test/test_helper.rb
# frozen_string_literal: true

$LOAD_PATH.unshift File.expand_path('../lib', __dir__)
require 'jp_vaccination'

require 'minitest/autorun'


test/vaccination_test.rb
# frozen_string_literal: true

require_relative './test_helper'
require_relative '../lib/jp_vaccination'
require_relative '../lib/jp_vaccination/vaccination'

class JpVaccination::VaccinationTest < Minitest::Test
  def test_vaccination_key_exists
    vaccination = JpVaccination.find('rotavirus_2')
    assert_equal vaccination.name, 'ロタウイルス'
    assert_equal vaccination.period, '２回目'
    assert_equal vaccination.regular, true
    assert_nil vaccination.type
    assert_equal vaccination.recommended, { "month": 3 }
    assert_equal vaccination.deadline, {
      "date_type": 'day',
      "start": nil,
      "end": 168,
      "less_than": true
    }
    assert_equal vaccination.interval, {
      "date_type": 'week',
      "start": 4,
      "end": nil,
      "less_than": true
    }
    assert_equal vaccination.formal_name, 'ロタウイルス ２回目'
  end

  def test_formal_name
    vaccination = JpVaccination.find('mumps_1')
    assert_equal vaccination.formal_name, 'おたふくかぜ １回目'
  end
end


test/jp_vaccination_test.rb
# frozen_string_literal: true

require_relative './test_helper'
require_relative '../lib/jp_vaccination'

class JpVaccinationTest < Minitest::Test # rubocop:disable Metrics/ClassLength
  def test_find_when_argument_is_not_exist_key
    not_exist_key = 'hib_5'
    e = assert_raises ArgumentError do
      JpVaccination.find(not_exist_key)
    end
    assert_equal 'The vaccination_key doesn\'t exist.', e.message
  end

  def test_next_day_when_interval_nil
    vaccination = 'hib_1'
    birthday = '2020-01-01'
    next_day = JpVaccination.next_day(vaccination_key: vaccination, last_time: birthday)
    assert_equal 'ヒブ １回目', next_day[:name]
    assert_equal Date.parse('2020-03-01'), next_day[:date]
  end

  def test_next_day_when_interval
    vaccination = 'hepatitis_B_2'
    last_time = '2020-01-01'
    next_day = JpVaccination.next_day(vaccination_key: vaccination, last_time: last_time)
    assert_equal 'Ｂ型肝炎 ２回目', next_day[:name]
    assert_equal Date.parse('2020-01-28'), next_day[:date]
  end

  def test_next_day_method_when_argument_is_not_exist_key
    not_exist_key = 'hib_5'
    e = assert_raises ArgumentError do
      JpVaccination.next_day(vaccination_key: not_exist_key, last_time: '2020-01-02')
    end
    assert_equal 'The vaccination_key doesn\'t exist.', e.message
  end

  def test_calc_date_when_day
    vaccination_period = JpVaccination.json_data['pneumococcus_2'.to_sym][:interval]
    last_time = '2020-01-01'
    calc_date = JpVaccination.calc_date(period: vaccination_period, start_or_end: :start, date: last_time)
    assert_equal Date.parse('2020-01-28'), calc_date
  end

  def test_calc_date_when_week
    vaccination_period = JpVaccination.json_data['rotavirus_2'.to_sym][:interval]
    last_time = '2020-01-01'
    calc_date = JpVaccination.calc_date(period: vaccination_period, start_or_end: :start, date: last_time)
    assert_equal Date.parse('2020-01-29'), calc_date
  end

  def test_calc_date_when_month
    vaccination_period = JpVaccination.json_data['Japanese_encephalitis_3'.to_sym][:interval]
    last_time = '2020-01-01'
    calc_date = JpVaccination.calc_date(period: vaccination_period, start_or_end: :start, date: last_time)
    assert_equal Date.parse('2020-07-01'), calc_date
  end

  def test_calc_date_when_year
    vaccination_period = JpVaccination.json_data['MR_1'.to_sym][:deadline]
    birthday = '2020-01-01'
    calc_date = JpVaccination.calc_date(period: vaccination_period, start_or_end: :start, date: birthday)
    assert_equal Date.parse('2021-01-01'), calc_date
  end

  def test_recommended_schedules # rubocop:disable Metrics/MethodLength
    birthday = '2021-04-01'
    expect_schedules = [{ name: 'ヒブ １回目', date: Date.parse('2021-06-01') },
                        { name: 'ヒブ ２回目', date: Date.parse('2021-07-01') },
                        { name: 'ヒブ ３回目', date: Date.parse('2021-08-01') },
                        { name: 'ヒブ ４回目', date: Date.parse('2022-04-01') },
                        { name: 'Ｂ型肝炎 １回目', date: Date.parse('2021-06-01') },
                        { name: 'Ｂ型肝炎 ２回目', date: Date.parse('2021-07-01') },
                        { name: 'Ｂ型肝炎 ３回目', date: Date.parse('2021-11-01') },
                        { name: 'ロタウイルス １回目', date: Date.parse('2021-06-01') },
                        { name: 'ロタウイルス ２回目', date: Date.parse('2021-07-01') },
                        { name: 'ロタウイルス ３回目', date: Date.parse('2021-08-01') },
                        { name: '小児用肺炎球菌 １回目', date: Date.parse('2021-06-01') },
                        { name: '小児用肺炎球菌 ２回目', date: Date.parse('2021-07-01') },
                        { name: '小児用肺炎球菌 ３回目', date: Date.parse('2021-08-01') },
                        { name: '小児用肺炎球菌 ４回目', date: Date.parse('2022-04-01') },
                        { name: '４種混合 第１期 １回目', date: Date.parse('2021-07-01') },
                        { name: '４種混合 第１期 ２回目', date: Date.parse('2021-08-01') },
                        { name: '４種混合 第１期 ３回目', date: Date.parse('2021-09-01') },
                        { name: '４種混合 第１期 ４回目', date: Date.parse('2022-04-01') },
                        { name: '２種混合 第２期', date: Date.parse('2032-04-01') },
                        { name: 'ＢＣＧ ', date: Date.parse('2021-09-01') },
                        { name: '麻しん・風しん混合 第１期', date: Date.parse('2022-04-01') },
                        { name: '麻しん・風しん混合 第２期', date: Date.parse('2026-04-01')..Date.parse('2027-03-31') },
                        { name: '水痘 １回目', date: Date.parse('2022-04-01') },
                        { name: '水痘 ２回目', date: Date.parse('2022-10-01') },
                        { name: 'おたふくかぜ １回目', date: Date.parse('2022-04-01') },
                        { name: 'おたふくかぜ ２回目', date: Date.parse('2026-04-01')..Date.parse('2027-03-31') },
                        { name: '日本脳炎 第１期 １回目', date: Date.parse('2024-04-01') },
                        { name: '日本脳炎 第１期 ２回目', date: Date.parse('2024-05-01') },
                        { name: '日本脳炎 第１期 ３回目', date: Date.parse('2025-04-01') },
                        { name: '日本脳炎 第２期', date: Date.parse('2030-04-01') }]
    assert_equal expect_schedules, JpVaccination.recommended_schedules(birthday)
  end

  def test_pre_school_year_born_april_1st
    birthday = '2020-04-01'
    assert_equal Date.parse('2025-04-01')..Date.parse('2026-03-31'), JpVaccination.pre_school_year(birthday)
  end

  def test_pre_school_year_born_april_2nd
    birthday = '2020-04-02'
    assert_equal '2026-04-01〜2027-03-31', JpVaccination.pre_school_year(birthday, true)
  end

  def test_sort_recommended_schedules
    birthday = '2022-04-14'
    sort_schedules = { '2022-06-14' => ['ヒブ １回目', 'ロタウイルス １回目', '小児用肺炎球菌 １回目', 'Ｂ型肝炎 １回目'],
                       '2022-07-14' => ['ヒブ ２回目', 'ロタウイルス ２回目', '小児用肺炎球菌 ２回目', '４種混合 第１期 １回目', 'Ｂ型肝炎 ２回目'],
                       '2022-08-14' => ['ヒブ ３回目', 'ロタウイルス ３回目', '小児用肺炎球菌 ３回目', '４種混合 第１期 ２回目'],
                       '2022-09-14' => ['４種混合 第１期 ３回目', 'ＢＣＧ '],
                       '2022-11-14' => ['Ｂ型肝炎 ３回目'],
                       '2023-04-14' => ['おたふくかぜ １回目', 'ヒブ ４回目', '小児用肺炎球菌 ４回目', '水痘 １回目', '麻しん・風しん混合 第１期',
                                        '４種混合 第１期 ４回目'],
                       '2023-10-14' => ['水痘 ２回目'],
                       '2025-04-14' => ['日本脳炎 第１期 １回目'],
                       '2025-05-14' => ['日本脳炎 第１期 ２回目'],
                       '2026-04-14' => ['日本脳炎 第１期 ３回目'],
                       '2028-04-01〜2029-03-31' => ['おたふくかぜ ２回目', '麻しん・風しん混合 第２期'],
                       '2031-04-14' => ['日本脳炎 第２期'],
                       '2033-04-14' => ['２種混合 第２期'] }
    assert_equal sort_schedules, JpVaccination.sort_recommended_schedules(birthday, true)
  end

  def test_convert_to_strings
    birthday = '2020-02-29'

    sort_schedules = { Date.parse('2020-04-29') => ['ヒブ １回目', 'ロタウイルス １回目', '小児用肺炎球菌 １回目', 'Ｂ型肝炎 １回目'],
                       Date.parse('2020-05-29') => ['ヒブ ２回目', 'ロタウイルス ２回目', '小児用肺炎球菌 ２回目', '４種混合 第１期 １回目',
                                                    'Ｂ型肝炎 ２回目'],
                       Date.parse('2020-06-29') => ['ヒブ ３回目', 'ロタウイルス ３回目', '小児用肺炎球菌 ３回目', '４種混合 第１期 ２回目'],
                       Date.parse('2020-07-29') => ['４種混合 第１期 ３回目', 'ＢＣＧ '],
                       Date.parse('2020-09-29') => ['Ｂ型肝炎 ３回目'],
                       Date.parse('2021-02-28') => ['おたふくかぜ １回目', 'ヒブ ４回目', '小児用肺炎球菌 ４回目', '水痘 １回目', '麻しん・風しん混合 第１期',
                                                    '４種混合 第１期 ４回目'],
                       Date.parse('2021-08-29') => ['水痘 ２回目'],
                       Date.parse('2023-02-28') => ['日本脳炎 第１期 １回目'],
                       Date.parse('2023-03-29') => ['日本脳炎 第１期 ２回目'],
                       Date.parse('2024-02-29') => ['日本脳炎 第１期 ３回目'],
                       Date.parse('2025-04-01')..Date.parse('2026-03-31') => ['おたふくかぜ ２回目', '麻しん・風しん混合 第２期'],
                       Date.parse('2029-02-28') => ['日本脳炎 第２期'],
                       Date.parse('2031-02-28') => ['２種混合 第２期'] }
    assert_equal sort_schedules, JpVaccination.sort_recommended_schedules(birthday)
  end
end


lib/jp_vaccination.rb
# frozen_string_literal: true

require 'date'
require 'json'
require_relative './jp_vaccination/vaccination'
require_relative './jp_vaccination/version'

module JpVaccination
  class << self
    def json_data
      vaccinations = File.open('./data/vaccinations.json', 'r', &:read)
      JSON.parse(vaccinations, symbolize_names: true)[:vaccinations]
    end

    def find(vaccination)
      data = json_data[vaccination.to_sym]
      output_argument_error(data)
      JpVaccination::Vaccination.new(data)
    end

    def recommended_schedules(birthday, convert_to_strings = nil)
      json_data.map do |_key, vaccination|
        name = "#{vaccination[:name]} #{vaccination[:period]}"

        if vaccination[:recommended].class != Hash
          recommended_day = pre_school_year(birthday, convert_to_strings)
        elsif vaccination[:recommended][:month]
          recommended_day = Date.parse(birthday) >> vaccination[:recommended][:month].to_i
        elsif vaccination[:recommended][:year]
          recommended_day = Date.parse(birthday) >> vaccination[:recommended][:year].to_i * 12
        end
        if convert_to_strings && recommended_day.instance_of?(Date)
          recommended_day = recommended_day.strftime('%Y-%m-%d')
        end
        { name: name, date: recommended_day }
      end
    end

    def sort_recommended_schedules(birthday, convert_to_strings = nil)
      ary =
        recommended_schedules(birthday, convert_to_strings).map do |key, _value|
          { key[:name] => key[:date] }
        end

      flatten_hash = {}.merge(*ary)
      ary_sort_date = flatten_hash.sort_by do |_name, date|
        date.instance_of?(Date) || date.instance_of?(String) ? date : date.first
      end
      ary_sort_date.to_h.each_key.group_by { |date| flatten_hash[date] }.each_pair { |_date, name_ary| name_ary.sort! }
    end

    def next_day(vaccination_key:, last_time:)
      next_day = {}
      json_data.each do |key, vaccination|
        next unless key == vaccination_key.to_sym

        name = "#{vaccination[:name]} #{vaccination[:period]}"

        date = case vaccination[:interval]
               when nil
                 calc_date(period: vaccination[:deadline], start_or_end: :start, date: last_time)
               else
                 calc_date(period: vaccination[:interval], start_or_end: :start, date: last_time)
               end
        next_day[:name] = name
        next_day[:date] = date
      end
      output_argument_error(next_day[:name])
      next_day
    end

    def calc_date(period:, start_or_end:, date:)
      date = case period[:date_type]
             when 'day'
               Date.parse(date) + period[start_or_end].to_i
             when 'week'
               Date.parse(date) + (period[start_or_end].to_i * 7)
             when 'month'
               Date.parse(date) >> period[start_or_end].to_i
             when 'year'
               Date.parse(date) >> (period[start_or_end].to_i * 12)
             end
      return date if start_or_end != :end

      period[:less_than] ? date - 1 : date
    end

    def pre_school_year(birthday, convert_to_strings = nil)
      fifth_birthday = Date.parse(birthday) >> 12 * 5
      case fifth_birthday.month
      when 1..3
        year = fifth_birthday.year
      when 4
        year = fifth_birthday.day == 1 ? fifth_birthday.year : fifth_birthday.year.next
      when 5..12
        year = fifth_birthday.year.next
      end
      convert_to_strings ? "#{year}-04-01〜#{year.next}-03-31" : Date.new(year, 4, 1)..Date.new(year.next, 3, 31)
    end

    def output_argument_error(not_exist_key)
      raise ArgumentError, 'The vaccination_key doesn\'t exist.' if not_exist_key.nil?
    end
  end
end
